File: solutions/ruby/perfect-numbers/1/perfect_numbers.rb
class PerfectNumber
    def self.classify(n)
        raise RuntimeError if n <= 0

        sum_of_factors = factors_of(n).sum

        case 
        when sum_of_factors == n
            "perfect"
        when sum_of_factors < n
            "deficient"
        else
            "abundant"
        end
    end

    private
    def self.factors_of(n)
        (1..n / 2).filter { | i | n % i == 0 }
    end
end

File: solutions/lua/perfect-numbers/1/perfect-numbers.lua
local function aliquot_sum(n)
    sum = 0
    for i=1,n/2 do 
        if n % i == 0 then
            sum = sum + i
        end
    end

    return sum
end

local function classify(n)
    sum = aliquot_sum(n)
    if sum == n then
        return "perfect"
    elseif sum < n then
        return "deficient"
    else 
        return "abundant"
    end
end

return { aliquot_sum = aliquot_sum, classify = classify }
